### DISCUSSION TITLE
Component Request: Device (For over-powered responsiveness)

### DISCUSSION THREAD

**Discussion by a-y-u-s-h** [chakra-ui/chakra-ui#1990] [Ideas]
# Feature

## Component: Device

One of the biggest issue with solution to responsiveness w/ Chakra UI is that solution isn't component based. From what I know after reading docs, media queries are currently dealt with by passing in props, which solves one part of the problem with responsive design - the dimensions. However it doesn't allow complete flexibility. Like if I want to render X component on laptops and Y component on tablets - I can't do that with this alone, because props will only change certain styles, not whole components (example: if I want a grid on laptops but 2 different components, one being canvas on mobile for some random reason - I can't do that) - I would need some ugly conditional logic mixed with the code. A common use case for this is implementation of Hamburger menu. These kind of issues can be solved gracefully with one additional component.

## Describe the solution you'd like
I would love to have a component (called `Device`) that renders childrens based on the screen size that they should be rendered on as props (optional, so whatever this component encapsulates gets rendered only if screen size is according to props). I have figured out one way to implement such component (implentation is towards the end of this issue). Before describing my own implementation that, let me give a taste of the usage and importance of the component.

Once the component is implemented, this is how `Device` component is supposed to feel like in code. `Device`, in code will render the children only in browser windows whose dimensions lie under the range that the props `represent`. `desktop` `tablet` `mobile` and a few more are valid props for this component.

1) Desktop (this piece of code renders `p` only on desktops and nothing on Mobiles or Tablets) -
 
```jsx
<Device desktop>
 <p>Hello</p>
</Device>
```

2) Mobile (this renders nothing on Desktop or Tablets, and renders a `p` only on devices matching `mobile` size) -
 
```jsx
<Device mobile>
 <p>Hello</p>
</Device>
```

3) Tablet (similar to above, render nothing on Desktop and Mobile - renders its children only on tablet sized browser window) -
 
```jsx
<Device tablet>
 <p>Hello</p>
</Device>
```

4) Mobile and Desktops (renders on desktop and mobile but not on Tablets) -
 
```jsx
<Device desktop mobile>
 <p>Hello</p>
</Device>
```

And likewise, these props can be supplied all three at a time or two at a time, and the component must render based on screen sizes. An extra `not` boolean prop should reverse the final render condition by other props (so if props are: `not desktop mobile`, children will not be rendered on `desktop` and `mobile` screen sizes). Useful when you want to keep the markup clean. Consider this code ("Hello from desktop" is rendered only on desktops and "Hello from Tablets or Mobile phones!" on every other screen size):

```jsx
<Device desktop>
 <p>Hello from Desktop</p>
</Device>
<Device not desktop>
 <p>Hello from Tablets or Mobile phones!</p>
</Device>
```

Also, for a finer control of responsiveness, `desktop`, `mobile` and `tablet` props can be either `boolean` or `string`. So that the usage becomes much better, like say there's a component `X` that must be rendered only on Medium & Large Desktops and Small mobile phones, the developer can write the following code: 

```jsx
<Device desktop="M L" mobile="S">
 <p>This component will be rendered only on M, L size desktops and S sized mobile phones!</p>
</Device>
```

## Describe alternatives you've considered

I used `react-responsive` library to create the component that behaves the exact way I've described above. The implementation looks like this (Code may look very cringy but it works fine, it's not refactored yet):

```jsx
import { useMediaQuery } from "react-responsive"

export const Device = ({ children, desktop, mobile, tablet, not }) => {
  /*
    ======================================
      This component renders children
      based on size of the screen. The usage
      is extremely simple but the component
      definition looks kind of ugly.
    ======================================
  */
  let condition = false
  const validate = (variable, name) => {
    /*
      ======================================
        This function basically takes in
        the variable and the name of the variable
        (should lie as a key in 'range' variable below).
        And then implemnts all the logic required to set
        the final condition.
      ======================================
    */
    const range = {
      mobile: {
        S: { maxWidth: 320 },
        M: { minWidth: 321, maxWidth: 375 },
        L: { minWidth: 376, maxWidth: 425 },
        XL: { minWidth: 426, maxWidth: 767 },
        generic: { maxWidth: 767 }
      },
      tablet: {
        generic: { minWidth: 768, maxWidth: 991 }
      },
      desktop: {
        S: { minWidth: 992, maxWidth: 1024 },
        M: { minWidth: 1024, maxWidth: 1360 },
        L: { minWidth: 1360, maxWidth: 2000 },
        XL: { minWidth: 2000 },
        generic: { minWidth: 992 }
      }
    }
    /*
      ======================================
        If that variable is a stirng, then
        each word in that string must be a valid
        size. And if that's true then we include the
        contribution of condition corresponding to that
        size in the final result. If not, we simply ignore it.
      ======================================
    */
    if (variable) {
      if (typeof variable === "string" || variable instanceof String) {
        variable.split(" ").forEach(i => {
          const size = range[name] ? (range[name][i] ? useMediaQuery(range[name][i]) : undefined) : undefined
          if (typeof size !== "undefined") condition = condition || size
        })
      } else if (typeof variable === "boolean") {
        if (range[name]) condition = condition || useMediaQuery(range[name].generic)
      }
    }
  }
  /*
    ======================================
      Finally, we run the checks and also
      if user doesn't provide any device,
      by default its assumed that the children
      should be rendered on all screens.
    ======================================
  */
  validate(mobile, "mobile")
  validate(tablet, "tablet")
  validate(desktop, "desktop")
  if (!desktop && !mobile && !tablet) condition = true
  condition = not ? !condition : condition
  return condition ? children : null
}

```

## Additional Context

**Is there a possibility that markup becomes cluttered if you have to wrap every component in a Device?**

The answer is 'NO!' :), the use of this `Device` component is optional. This makes things very clean (compared to the usual media query alternative - for which developers will have to remember the screen sizes). You'd use it only when you want to render something different for different screen sizes - which means the developers can keep using Chakra UI like they're doing right now, and use Device when they feel like it (for bigger components, not buttons or smaller components). Like let's say you need to have a layout that becomes a 2 column layout on a mobile and a 3 column layout on a computer - you'd do this:

```jsx

export const Layout = ({ children }) => (
<Device desktop>
  <Box display="grid" gridTemplateColumns="1fr 1fr 1fr">
    {children}
  </Box>
</Device>
<Device not desktop>
  <Box display="grid" gridTemplateColumns="1fr 1fr">
    {children}
  </Box>
</Device>
)
```

Notice how this also allows the existence of some crazy useful components. Like the above component can be modified to this:

```jsx
export const Layout = ({ children, desktop, tabloids }) => (
<Device desktop>
  <Box display="grid" gridTemplateColumns={desktop}>
    {children}
  </Box>
</Device>
<Device not desktop>
  <Box display="grid" gridTemplateColumns={tabloids}>
    {children}
  </Box>
</Device>
)

```

And now you get a generic Layout component which gives you different grids layouts for different devices:

```jsx
<Layout desktop="1fr 1fr 1fr" tabloids="1fr 1fr">
...
</Layout>

<Layout desktop="1fr 3fr 2fr" tabloids="1fr">
...
</Layout>
```

The `not` prop is very useful in keeping the markup clean. If you pass in the `desktop` and `tablet` prop as boolean, children will be rendered only on desktops and tablets (not mobile). If you pass in `not` along with it - the final conditional will be reversed - meaning children will render on all screens that are `not` desktop or tablet - which means `mobile`. The rest of the library remains same, no difference in development styles except this additional component will probably be the best solution for responsiveness in React world. It could even make further responsive components possible. Like the ones [Tailwind UI](https://www.tailwindui.com/) comes with.

Another example:

```jsx
<p>Hello on every device</p>

<Device not desktop>
     <p>Hello only from Mobiles and Tablets!</p>
</Device>

```

Above code will render The "Hello on every device" on all screen sizes, and "Hello only from Mobiles and Tablets" paragraph will not get rendered on Computers, will get rendered only on Tablets and Mobiles. Sizes according to this object (more can be added):

```jsx
{
      mobile: {
        S: { maxWidth: 320 },
        M: { minWidth: 321, maxWidth: 375 },
        L: { minWidth: 376, maxWidth: 425 },
        XL: { minWidth: 426, maxWidth: 767 },
        generic: { maxWidth: 767 }
      },
      tablet: {
        generic: { minWidth: 768, maxWidth: 991 }
      },
      desktop: {
        S: { minWidth: 992, maxWidth: 1024 },
        M: { minWidth: 1024, maxWidth: 1360 },
        L: { minWidth: 1360, maxWidth: 2000 },
        XL: { minWidth: 2000 },
        generic: { minWidth: 992 }
      }
    }
```

Like, to render on every screen size, you just simply don't use the `Device` component or you use it by passing it all three props `tablet` `mobile` `desktop` or by passing no props at all (in the implementation, the third last line makes the children render on every device if none of tablet, mobile or desktop props are passed). That said, it probably will be the 2nd most used component after `Box`. This functionality could also be added inside `Box` itself but semantically `Device` makes more sense. Imagine calling `<Box computer>` - I mean yeah, both are cool and both would enable a lot of clean responsive code, but keeping this reponsive functionality out will not involve extra over-head in more used Box components.

Even crazier components are possible, like you can have a component in X place inside 2nd block of the grid on desktop, but on Mobile you can put it in 1st block between the components that already exist inside 1st block! Here's an example outcome (video appears to have come between heading and paragraphs on mobile): (Device components used on Evergreen UI):

![ezgif-4-2f124e4727ae](https://user-images.githubusercontent.com/19586719/85213234-2a480880-b379-11ea-8906-72c451a5166d.gif)

**Comment by with-heart** [comment]
Hi @a-y-u-s-h!

First, thank you for the _extremely_ detailed component request! I appreciate the time and effort you put into this. You provided a thorough explanation, multiple examples, sample code, and the answer to just about any question I'd have to ask. It makes our job as maintainers much easier to have ideas presented to us in this way. ❤️

I love the willingness to explore a thoughtful new API, but my main concern with this feature is complexity:

- Adds a second method for handling responsiveness to Chakra

Almost all of the _style props_ in the Chakra system take responsive array and object values. While they may not allow the ultimate flexibility in breakpoint display management that your implementation does, they all work exactly the same. Another method of handling the same thing means more confused users and more documentation.

- Complex props combinations

The combination of `boolean` size props, `string` size props, and the `not` prop allows for a large number of possible results. This increases the cognitive overhead of understanding at-a-glance what each use of the `Device` component is doing. While I imagine this wouldn't be an issue for most users who don't need that level of detail, having been at this awhile, I know that some users have a tendency to get themselves lost in overly-complex features.

- Far more detailed breakpoints

Up to this point, we've not had a request that involved expanding our breakpoints beyond the simple array of `sm`, `md`, `lg`, and `xl` that we user currently. Adding fine-grained control not only of additional breakpoints but nested breakpoints would require a serious overhaul of the system, not to mention that they'd conflict with the existing array method of using responsive props in Chakra.

---

I would love some feedback from yourself, other users, and other maintainers, but for these reasons, I'm hesitant to add something like this to Chakra.

With that said, your request does have me thinking about some other ideas:

- I really love the way props work for the `Layout` component and would like to figure out a way to simplify props in that manner. It's much easier to process props that have been individualized to a single breakpoint, rather than trying to specify them under a single prop, e.g. `<Grid gridTemplateColumns={["1fr", "1fr 1fr", "1fr 1fr 1fr", "1fr 1fr 1fr 1fr"]} />`.

- Add a `Hidden` component to allow for selectively rendering the contents

This component would have the `above` and `below` props, allowing you to conditionally render content based on breakpoint names:

```jsx
<Hidden above="lg">
  I'm displayed on any device except mobile
</Hidden>

<Hidden below="lg">
  I'm displayed only on desktop
</Hidden>

<Hidden above="sm">
  I'm displayed only on mobile
</Hidden>
```

This would allow users to more easily display content only for specific breakpoints, while sticking with our existing breakpoints system and without as much of the complexity I mentioned above.

---

PS: This is totally unrelated, but I love your desktop background! Too bad we can't all live in a place like that.

**Reply by tcolinpa** [reply]
A Hidden component as described would be great!

**Reply by tcolinpa** [reply]
`<Box display={{ base: 'none', md: 'block' }}> `

This worked like a charm.

**Reply by pianomansam** [reply]
I recall a `<Hidden>` component in Material UI that worked a similar way and found myself needing that sort of functionality in Chakra. Here's a quick version that seems to be working well. 

https://gist.github.com/pianomansam/8228de4458af595add1fb45cec0cec02

It follows the suggested API described above:

```
<Hidden above="lg">
  I'm displayed on any device except mobile
</Hidden>

<Hidden below="lg">
  I'm displayed only on desktop
</Hidden>

<Hidden above="sm">
  I'm displayed only on mobile
</Hidden>
```

But it has several other additions:

By default it uses `display: none` which means the children elements are still in the DOM, just hidden by CSS.

If you want the children completely removed from the DOM, this supports doing so via a `remove` property.

```
<Hidden above="lg" remove>
  I'm displayed on any device except mobile
</Hidden>
```

This supports visually hidden children via a `visually` property, keeping them in the DOM and visible to screen readers. It does this by wrapping the children with `<VisuallyHidden>`
```
<Hidden above="lg" visually>
  I'm displayed on any device except mobile
</Hidden>
```

**Comment by a-y-u-s-h** [comment]
@with-heart Yes, actually I just wanted to suggest this component based solution for responsiveness so that maintainers can adapt it with Chakra UI conventions like you suggested (`sm`, `md`, `lg`, etc) or as an alternative utility component (since it doesn't do anything with styles, it can't be a part of theme-ing system). The `Hidden` component that you mentioned is very similar to the notation I used:

```js
<Hidden above="lg">
I'm displayed on any device except mobile.
</Hidden>
```

So, assuming `lg` is `desktop` and `md` is `tablet` and `sm` is `mobile`, above code is equivalent to `Device` with a `not` prop set to true:

```js
<Device not mobile>
I'm displayed on any device except mobile
</Device>

or 

<Device tablet desktop>
I'm displayed on any device except mobile
</Device>
```

The primary difference being that Device says "I will render the childrens inside me based on the props, if no props are passed I will assume you're OK with me rendering everywhere" and Hidden says "I will hide my children below or above the screen size you supply as props"...both cool!

It's all a matter of preferences really I think. I picked those prop naming convention and fine grained breakpoints because Chrome Developer tools uses those same breakpoints and names so while debugging it becomes very easy to see what renders on which screens (and it's easy to add new breakpoints or remove some). I agree whole heartedly that these combination of different props (and same prop being both a boolean or a string) can be confusing at first, but I in my experience it semantically gave me a good idea about what the code does after a while.

Device component's implementation can be modified to suit Chakra UI. Usage will look like this:

```js
<Device sm lg>
This renders only on sm and lg screens, not on mid sized screens.
</Device>
```

Here (if it's similar to my implementation) `lg` prop would mean the same thing as `above="lg"` (including `lg` range)

**Comment by TheThirdRace** [comment]
I think it's a bit over-complicated since we already have a hook for this that allows pretty much any combination:

```
import { Box, Button, useBreakpointValue } from '@chakra-ui/core'
export const MyComponent = (): ReactElement => {
	const mobileOrDesktop = useBreakpointValue({ base: <Box>On mobile</Box>, md: <Button>On desktop</Button> })
	return mobileOrDesktop
}
```

And if you want to hide the code because you find it ugly, simply create an HoC out of this logic.

**Reply by FrancoisCosta** [reply]
@TheThirdRace  I tried this and got 
`Error: Reader(...): Nothing was returned from render. This usually means a return statement is missing. Or, to render nothing, return null.`
any idea why? are you sure this works?
Thanks

**Reply by TheThirdRace** [reply]
You tried this "as-is"?

Because this example is for Chakra < 1.0, when `@chakra-ui/core` was still a thing...

Nowadays, you need to use `@chakra-ui/react`

**Reply by FrancoisCosta** [reply]
hey, thanks for the quick response :) 

I tried it with `@chakra-ui/react` not the old version.
Apart from that, I kept the rest as is except I removed the `: ReactElement` as I am not using typescript

**Reply by Daim-Nickel-Penny** [reply]
Thanks!

**Comment by TheThirdRace** [comment]
To add to my previous post...

If you don't mind having stuff not displayed in the DOM, you can also use the theming options and overwrite **display**:

```
export const MyComponentStyle = {
	parts: ['header', 'body', 'footer'],
	baseStyle: {
		header: {
			display: 'flex',
			flexWrap: 'nowrap',
			justifyContent: { base: 'space-between', md: 'normal' }
		},
		body: {
			display: { base: 'none', md: 'flex' },
			flexGrow: '2',
		},
		footer: { display: { base: 'flex', md: 'none' } }
	},
	defaultProps: {}
}
```

Don't pay too much attention to the actual values, I did this in a hurry, but you can see I use **display** here to hide some sections according to breakpoint values.

As to why someone would render it and then hide it (especially while using React)?

Well, it always depends on context.

Working on my own website, I used the **display:none** method in my page template because it allowed me to shave 12KB from the bundle size. I know the DOM is a bit bigger, but if it allows me to execute faster and shave 100ms for the user, I guess it's not such a bad compromise.

Most people try to do as much as possible in React to avoid using basic HTML or CSS. I would argue displaying a section or not is a layout issue and layout should be CSS. Obviously, there are situations where it's more advisable to do the opposite too. It's a matter of context.

It's all about compromises. If you're doing half the layout in React and the other half in CSS, you're gonna have a wonderful time to maintain it later... 

There have been some challenges though, but it's expected given Chakra-UI 1.0 is still not out yet.

**Comment by a-y-u-s-h** [comment]
@TheThirdRace Actually yeah I would agree that `display: none` has its place. I use it too in a lot of places. It's just that -

1)  The composition API enables a lot of other cases of responsiveness when you want to have a totally different markup depending on the screen-size (and have a cleaner code - without manually attaching classes and hiding them yourself, you just specify the props - named breakpoints, which are the same as the ones that devtools support).

2) There are cases when you **don't** want to render something at all, instead of hiding it to get performance especially when you're working with a lot of SVG or Canvas elements.

3) I like the idea of having all layout related things in CSS, except I feel that `render only on certain screen sizes` is a solution that can only be solved by a framework (like React). It's like leveraging abilities of framework to achieve performance gains. The layout is still being managed by specified styles but what to render is decided by a component.   

Currently, like working with ThemeUI - I have variants for things that hide upon hitting certain breakpoints, because they do the job, and this custom component to avoid rendering things that don't need to be rendered on small screens at all. Though I feel now this component that renders on different screen sizes is better if it's a different package in itself, that way Chakra will remain free from multiple ways of handling the same problem.

**Comment by TheThirdRace** [comment]
I understand your need and your arguments about using display property.

I will reiterate we already have **useBreakPointValue** hook and you can easily make your own HoC for it.

I wouldn't dare speak for Chakra UI's team, but my personal opinion is there are many other urgent matters that needs to be addressed before thinking about making an official HoC for this. Especially since making your own HoC would take less than an hour...

Some food for thoughts about CSS-In-Js... I love the fact you can really manage all your CSS needs as a "component", use composition of styles and not have to worry about critical CSS. But as soon as you write 1 style property on any component, it's not reusable in another project without you having to rewrite it.

Example that CANNOT be reused
```
<Box mr={4}>
```

Example that CAN be reused in any project
```
const styles = useStyleConfig('Box', {})
...
<Box sx={styles}
```

By abstracting all styling under CSS, you do make sure your components are 100% reusable elsewhere. All you have to do is override the theme and you're good to go.

I agree with you 100% that not all components should use **display** for rendering or not. It's a matter of compromises. Sometimes it's better to have a bigger DOM and less JS while other times it's better to have more JS and a smaller DOM. Each have their own usecase and it's our job to figure out which one is more efficient. Always keep in mind that 50KB of CSS is much, much, much faster than 50KB of JS. Not only for parsing, but also rendering.

**Reply by a-y-u-s-h** [reply]
> Some food for thoughts about CSS-In-Js... I love the fact you can really manage all your CSS needs as a "component", use composition of styles and not have to worry about critical CSS. But as soon as you write 1 style property on any component, it's not reusable in another project without you having to rewrite it.

> Example that CANNOT be reused

```jsx
<Box mr={4}>
```
---

Correction: This can also be used in any project. I usually keep styles in a collection of YAML files, which gets resolved as a JS object - so that I can use it in another framework as well. So if I have styles that look like this:

```yaml
box:
  marginRight: 4px
```

I can use it by directly passing in the props as well:

```jsx
import styles from "path-to-styles.yaml"

// Equivalent to passing in the props.

return <Box {...styles.box} />
```

Re: useBreakPointValue - it's a useful hook, agreed. But I suggest a component API instead because it looks cleaner. Also with a little more effort into the implementation, it should be able to accept generic values (min-size and max-size) because behind the scenes it uses a hook anyways and returns children conditionally. So it's sort of like a higher level abstraction.

**Reply by TheThirdRace** [reply]
That's a great example for the styles, never thought of doing it like that.

Although, your example also doesn't contradict my original point either.

My point was that writing a static styling prop on the component itself, ex: `<Box mr={4}>`, essentially make your component not reusable in another project. 

Your example did modify the static styling by importing the styles from another file and didn't use a direct styling props (mr, ml, display, etc.) on the component itself.

The example you provided gives a great alternative to the **useStyleConfig()** method. Thanks for sharing it.

**Reply by a-y-u-s-h** [reply]
Agreed. I usually avoid using hooks at all for styling, because hooks are inherently a pattern used by React - so it's framework dependent. All my styles are into YAML files to be used in Theme UI (components.yaml - contains reusable variants, foundations.yaml contains colors, fonts etc, and application.yaml contains page-specific variants). And I always use the verbose camel-cased keys so that my styles are framework independent by default. Theme UI gives too much advantage, so using sx prop makes sense.

But to make it really re-usable (framework independent, I think Theme UI doesn't work in with other frameworks), sx prop must be ditched as well and included in maybe a file that contains these static props for all components in markup. 

```jsx
<Box {...styles.container}>
  {children}
</Box>
```

In this case, If I'm using Theme UI or if I know I'm not going to convert this project ever into another framework, I can always save my static props for a particular page as:

```yaml
container:
  sx:
    width: 100%
    height: 100%
    background: gray.800
```

And if I want to have styles independent of Theme UI, since there's always a possibility that maybe the employer decides to shift everything to Vue, I would keep the `style` prop instead, and my markup will remain same. 

```yaml
container:
  style:
    width: 100%
    height: 100%
    background: "#242424"
```

But, it's some work and I avoid it during development (only do it after I'm done playing around with UI), although it does result in really nice separation of concerns.

**Comment by ljosberinn** [comment]
FYI: moved this to discussions as there is clearly something to be discussed here.

Generally speaking: a component with the initially proposed props such as
```js
<Device desktop>
 <p>Hello</p>
</Device>
```
would be hard, if not impossible, to realize given custom breakpoints:

```js
const breakpoints = createBreakpoints({
 xxs: '5rem',
 xs: '15rem',
 sm: '25rem',
 md: '35rem',
 lg: '45rem',
 xl: '55rem',
 xxl: '65rem',
 huge: '75rem',
})
```
Given those, ignoring the actual values, at which point would `desktop` kick in?

What I believe might work though is a hoc as proposed, using `useBreakpointValue` and props matching `breakpoints`:

```js
const normalizeBreakpoints = (breakpoints) => {
  return Object.entries(breakpoints).reduce((carry, [name, value]) => {
    if(value) {
      carry[name] = value === true ? 'initial' : value;
    } else {
      carry[name] = 'none'
    }

    return carry;
  }, {});
}

export function withBreakpoints(Component) {
  const C = ({xxs, xs, sm, md, lg, xl, xxl, huge, ,...rest}) => {
    const breakpoints = normalizeBreakpoints({ xxs, xs, sm, md, lg, xl, xxl, huge });
    const isVisible = useBreakpointValue(breakpoints);

    return isVisible ? <Component {...rest} /> : null;
  }

  C.displayName = `withBreakpoints(${component.displayName})`;

  return C;
}
```

just quickly scribbled. Thinking about it, it might work using `const { breakpoints } = useTheme()`. Needs investigation I guess, but I'm afraid doesn't have priority.

**Reply by a-y-u-s-h** [reply]
Yup, so the naming convention in the component I suggested above is based on Chrome Developer tools.  In the implementation you'd see max and min sizes for all screen sizes:

```jsx
    const range = {
      mobile: {
        S: { maxWidth: 320 },
        M: { minWidth: 321, maxWidth: 375 },
        L: { minWidth: 376, maxWidth: 425 },
        XL: { minWidth: 426, maxWidth: 767 },
        generic: { maxWidth: 767 }
      },
      tablet: {
        generic: { minWidth: 768, maxWidth: 991 }
      },
      desktop: {
        S: { minWidth: 992, maxWidth: 1024 },
        M: { minWidth: 1024, maxWidth: 1360 },
        L: { minWidth: 1360, maxWidth: 2000 },
        XL: { minWidth: 2000 },
        generic: { minWidth: 992 }
      }
    }
```

Developer, if they use chrome dev tools shouldn't need to remember these screen sizes. Also, the component provides finer control. Like if for some very weird reason I want to show the children on Desktop - Small and Large, but not Medium. I can simply write:

```jsx
<Device desktop="S L">...</Device>
```

And if I want to render something **not** on mobiles - M, L but on all other screen sizes including Mobile S and XL, I can do this:

```jsx
<Device not mobile="M L">...</Device>
```

All props other than `not` can be either booleans or strings. 

This takes away a lot of characters from the code compared to using a generic useBreakPointValue hook. And actually, with a few more lines of code, generic condition could be added as well in the implementation of device. I think actually this **Device**, since it's a layout __modifier__ component and doesn't really render anything on its own, it sort of deserves to be in its own package. I'll make one soon. This was a crude implementation.

**Comment by tasca** [comment]
Hi Ayush,

is your implementation ready?
can you give me?

**Reply by a-y-u-s-h** [reply]
One possible implementation is in the issue that I created (Alternatives that I've considered). Haven't turned it into an NPM package yet. Now I do responsive within the framework of tailwind but I agree this could be useful.
